server/oauth2.go

```diff
@@ -654,6 +654,71 @@ func validateRedirectURI(client storage.Client, redirectURI string) bool {
 	return err == nil && host == "localhost"
 }
 
+func isWildcardRedirectURIMatch(wildcardURI, redirectURI string) bool {
+	parsedWildcardURI, err := url.Parse(wildcardURI)
+	if err != nil {
+		return false
+	}
+
+	// Wildcard URIs must be https
+	if parsedWildcardURI.Scheme != "https" {
+		return false
+	}
+
+	// Wildcard URIs only apply to URIs with subdomains
+	wildcardDomains := strings.Split(parsedWildcardURI.Hostname(), ".")
+	if len(wildcardDomains) < 3 {
+		return false
+	}
+
+	// Wildcard URIs may only contain a single '*' and it must be in the lowest level domain
+	if strings.Count(wildcardURI, "*") != 1 {
+		return false
+	}
+	if !strings.Contains(wildcardDomains[0], "*") {
+		return false
+	}
+
+	parsedRedirectURI, err := url.Parse(redirectURI)
+	if err != nil {
+		return false
+	}
+	redirectDomains := strings.Split(parsedRedirectURI.Hostname(), ".")
+
+	return len(wildcardDomains) == len(redirectDomains) &&
+		wildcardMatch(wildcardDomains[0], redirectDomains[0]) &&
+		strings.Join(wildcardDomains[1:], ".") == strings.Join(redirectDomains[1:], ".") &&
```

Comment:
I think the first check `len` isn't required. The next two checks verify that and the `len` doesn't prevent indexing into slices of len < 2.

Reply:
Thanks - that was leftovers...removed.